### PR TITLE
Add local/writer.Events

### PR DESCRIPTION
**Pull request by michaljurecko**
Jira: https://keboola.atlassian.net/browse/PSGO-286

Changes:
- Added `local/writer.Writer` lifecycle events: `OnWriterOpen`, `OnWriterClose`.
  - It is used by the `statistics.Collector` in the next PR.
  
-----------------------

image

---------------------------

### CODE REVIEW

File: internal/pkg/service/buffer/storage/level/local/writer/base.go [hunk 1]
```diff
@@ -1,4 +1,4 @@
-package base
+package writer
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/storage/level/local/writer/base.go:1]: Mal som problem s cyklickou zavilostou medzi pkgs:

`writer (Factory) -> csv -> base -> EVENTS -> writer (Writer) `

Fixnute:
- `local/writer/base/base.go` :arrow_right:  `local/writer/base.go` 
- `local/writer/factory.go` :arrow_right:  `local/writer/factory/factory.go`
- ( :heavy_plus_sign: `local/writer/event.go` )

image

File: internal/pkg/service/buffer/storage/level/local/writer/base.go [hunk 2]
```diff
@@ -19,16 +19,18 @@ type syncer = disksync.Syncer
 type BaseWriter struct {
 	*chain
 	*syncer
+	events   *Events
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/storage/level/local/writer/base.go:22]: Potrebujem nejak napojit `statistics` na `local.Writer`, hlavne riesim, aby sa po `Writer.Close` ulozila posledna hodnota statistik do databazy. To sa potom pouziva na vyhodnotenie `conditions`.

Nechcel som vytvarat nejake priame spojenie medzi `local/writer` a `statistics` pkg.

Preto som pridal `writer.Events`: `OnWriterOpen`/`OnWriterClose`.

V tomto PR je iba pridanie tychto eventov + testy.

File: internal/pkg/service/buffer/storage/level/local/writer/event.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+// OnWriterClose registers a callback that is invoked in the LIFO order when a writer is closed.
+func (e *Events) OnWriterClose(fn func(Writer, error) error) {
+	e.onWriterClose = append(e.onWriterClose, fn)
+}
+
+// Clone returns a new independent instance of the Events with reference to the parent value.
+func (e *Events) Clone() *Events {
+	return &Events{parent: e}
+}
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/storage/level/local/writer/event.go:71]: Eventy je mozne definovat na `volumes`, `volume` a `writer` úrovni, na to sluzi `Clone`, vid testy.

File: internal/pkg/service/buffer/storage/level/local/writer/volume/volumes.go [hunk 4]
```diff
@@ -10,24 +10,31 @@ import (
...
 	*baseVolumes
+	events *writer.Events
 }
 
 // DetectVolumes function detects and opens all volumes in the path.
-func DetectVolumes(ctx context.Context, logger log.Logger, clock clock.Clock, path string, opts ...Option) (*Volumes, error) {
-	v, err := volume.DetectVolumes(logger, path, func(info volumeInfo) (*Volume, error) {
-		return Open(ctx, logger, clock, info, opts...)
+func DetectVolumes(ctx context.Context, logger log.Logger, clock clock.Clock, path string, opts ...Option) (out *Volumes, err error) {
+	events := writer.NewEvents()
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/storage/level/local/writer/volume/volumes.go:25]: Prazny zoznam registovanych event listeners sa vytvorí na `volumes` urovni, a posiela sa na `volume` uroven.

File: internal/pkg/service/buffer/storage/level/local/writer/volume/volume.go [hunk 5]
```diff
@@ -61,13 +63,14 @@ func NewInfo(path, typ, label string) volume.Info {
 //   - If the drainFile exists, then writing is prohibited and the function ends with an error.
 //   - The local.VolumeIDFile is loaded or generated, it contains storage.VolumeID, unique identifier of the volume.
 //   - The lockFile ensures only one opening of the volume for writing.
-func Open(ctx context.Context, logger log.Logger, clock clock.Clock, info volumeInfo, opts ...Option) (*Volume, error) {
+func Open(ctx context.Context, logger log.Logger, clock clock.Clock, events *writer.Events, info volumeInfo, opts ...Option) (*Volume, error) {
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/storage/level/local/writer/volume/volume.go:66]: `volume` uroven.

File: internal/pkg/service/buffer/storage/level/local/writer/volume/writer.go [hunk 6]
```diff
@@ -28,19 +28,28 @@ func (v *Volume) NewWriterFor(slice *storage.Slice) (w writer.Writer, err error)
 	// Setup logger
 	logger := v.logger
 
+	// Setup events
+	events := v.events.Clone()
+
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/storage/level/local/writer/volume/writer.go:33]: a posledna, `writer` uroven

File: internal/pkg/service/buffer/storage/level/local/writer/volume/writer.go [hunk 7]
```diff
@@ -94,14 +103,14 @@ func (v *Volume) NewWriterFor(slice *storage.Slice) (w writer.Writer, err error)
...
-
 	// Create writer via factory
-	return v.config.writerFactory(writer.NewBaseWriter(logger, v.clock, slice, dirPath, filePath, chain))
+	w, err := v.config.writerFactory(writer.NewBaseWriter(logger, v.clock, slice, dirPath, filePath, chain, events))
+	if err != nil {
+		return nil, err
+	}
+
+	// Wrap the writer to add events dispatching
+	return writer.NewEventWriter(w, events)
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/storage/level/local/writer/volume/writer.go:113]: konkretnu implementaciu `writer.Writer` (teda `csv.Writer`) obalime do `writer.EventWriter`.

File: internal/pkg/service/buffer/storage/level/local/writer/test/writer.go [hunk 8]
```diff
@@ -21,6 +21,7 @@ import (
 // Writer implements a simple writer implementing writer.Writer for tests.
 // The writer writes row values to one line as strings, separated by comma.
 type Writer struct {
+	helper  *WriterHelper
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/storage/level/local/writer/test/writer.go]: Pridanim `EventWriter` sa uz jednoducho nedostanem k `test.Writer.ExpectWritesCount/TriggerSync` co sa pouziva dost v testoch.

Radsej som to vyriesil lepsie, a tieto metody vyclenil do `WriterHelper`, a tede nie potrebne riesit "pretypovanie" (interface dereference) `writer.Writer -> *test.Writer`.

File: internal/pkg/service/buffer/storage/level/local/writer/volume/writer_test.go [hunk 9]
```diff
@@ -620,7 +621,7 @@ func (tc *writerTestCase) NewWriter(opts ...Option) (*test.Writer, error) {
 		return nil, err
 	}
 
-	return w.(*test.Writer), nil
+	return w, nil
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/storage/level/local/writer/volume/writer_test.go:624]: Toto uz nemusim robit.

File: internal/pkg/service/buffer/storage/level/local/writer/event_test.go [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+INFO  EVENT: slice: "2002-01-01T00:00:00.000Z", event: OPEN (2), level: volumes
+INFO  EVENT: slice: "2002-01-01T00:00:00.000Z", event: OPEN (1), level: volumes
+INFO  EVENT: slice: "2002-01-01T00:00:00.000Z", event: CLOSE (6), level: writer2
+INFO  EVENT: slice: "2002-01-01T00:00:00.000Z", event: CLOSE (5), level: writer2
+INFO  EVENT: slice: "2002-01-01T00:00:00.000Z", event: CLOSE (4), level: volume2
+INFO  EVENT: slice: "2002-01-01T00:00:00.000Z", event: CLOSE (3), level: volume2
+INFO  EVENT: slice: "2002-01-01T00:00:00.000Z", event: CLOSE (2), level: volumes
+INFO  EVENT: slice: "2002-01-01T00:00:00.000Z", event: CLOSE (1), level: volumes
+`), strings.TrimSpace(strhelper.FilterLines(`^INFO  EVENT: slice: "2002`, logger.InfoMessages())))
+}
```
Comment (michaljurecko) [inline internal/pkg/service/buffer/storage/level/local/writer/event_test.go:152]: Tu su doplnene testy k eventom.

**Review by zajca — APPROVED** [review]
Vypadá to, že to podle testů funguje :)

**Pull request merged**